inspect_dynspec/gpr_smooth.py

```diff
@@ -0,0 +1,266 @@
+
+
+def make_A_matvec(Ls : tuple, mask : Mask, prec_flat : jnp.ndarray) -> callable:
+    """
+    Returns a function A_matvec(z) that computes
+      (I + Lᵀ Rᵀ diag(prec_flat) R L) z
+    where R is the mask operator and Σ⁻¹ = diag(prec_flat).
+    Ls: list of matrices (e.g. [Lv, Lt])
+    """
+
+    @jax.jit
+    def A_matvec(z):
+        # 1) x = kron_mv(Ls, z)
+        x_flat = kron_mv(Ls, z)
+        # 2) Rx = mask.forward(x_flat)
+        Rx = mask.forward(x_flat)
+        # 3) Σ⁻¹ Rx  ← use per-pixel precisions at observed locs
+        prec_obs = mask.forward(prec_flat)
+        Sinv_Rx = Rx * prec_obs
+        # 4) Rᵀ (Σ⁻¹ Rx)
+        RT_Sinv_Rx = mask.adjoint(Sinv_Rx)
+        # 5) Lᵀ term: kron_mv([A.T for A in Ls[::-1]], RT_Sinv_Rx)
+        LT_term = kron_mv([A.T for A in Ls], RT_Sinv_Rx)
+        # 6) Return (I + …) z
+        return z + LT_term
+
+    return A_matvec
+
+
+def gpr_smooth(
+    data: jnp.ndarray,  # shape (Nv, Nt)
+    mask: jnp.ndarray,  # shape (Nv, Nt), bool or {0,1}
+    weights: Optional[
+        jnp.ndarray
+    ] = None,  # shape (Nv, Nt), per-pixel noise-precision (1/σ²)
+    l_length_nu: float = 1.0,
+    l_length_t: float = 1.0,
+    sigma2: float = 1.0,
+    jitter: float = 1e-6,
+    cg_tol: float = 1e-6,
+    cg_maxiter: int = 200,
+    nof_weight_samples: int = 20,
+) -> Tuple[jnp.ndarray, jnp.ndarray]:
+    """
+    2D Gaussian‐Process smoothing via an RBF kernel with optional heteroscedastic noise.
+
+    Args:
+      data     : float[H, W]  — observed brightness.
+      mask     : float[H, W]  — per-pixel mask (1 for valid pixels, 0 for invalid).
+      weights  : float[H, W]  — per-pixel noise-precision (1/sigma^2). If None, uniform weights are used.
+      l_length_nu : float     — RBF length-scale for frequency axis.
+      l_length_t : float      — RBF length-scale for time axis.
+      sigma2   : float        — RBF signal variance sigma^2.
+      jitter   : float        — small value added to diagonal for numerical stability.
+      cg_tol   : float        — CG solver tolerance.
+      cg_maxiter: int         — maximum number of CG iterations.
+      nof_weight_samples: int — number of latent samples to estimate posterior variance.
+
+    Returns:
+      float[H, W] — posterior mean (“smoothed”) image.
+      float[H, W] — posterior variance image.
+    """
+
+    # Prep:
+    Nv, Nt = data.shape
+    t_grid = jnp.linspace(0, 1, Nt, dtype=jnp.float64)
+    v_grid = jnp.linspace(0, 1, Nv, dtype=jnp.float64)
+    R = Mask(mask)
+
+    weights = jnp.ones_like(data) if weights is None else weights
+
+    data_flat = data.ravel()
+    prec_flat = weights.ravel()
+
+    y_obs = R.forward(data_flat)  # observed data
+    prec_obs = R.forward(prec_flat)  # observed precisions
+
+    # Σ = diag(1/prec_flat)  ⇒  σ_obs = 1/√prec_obs
+    sigma_obs = 1.0 / jnp.sqrt(jnp.where(prec_obs == 0, 1, prec_obs))
+
+    y_white = y_obs * prec_obs  # whitened observations
```

Comment:
Technically a whitened variable is normalised by its standard deviation (not variance) so calling `y_white` whitened is not really correct. Maybe weighted?